src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java

```diff
@@ -156,56 +160,191 @@ public PaginatedResult getComponents(final Project project, final boolean includ
+                        "I0"."REPOSITORY_URL" AS "integrityRepoUrl",
+                        COUNT(*) OVER() AS "totalCount"
+                FROM "COMPONENT" "A0"
+                INNER JOIN "PROJECT" "B0" ON "A0"."PROJECT_ID" = "B0"."ID"
+                LEFT JOIN "INTEGRITY_META_COMPONENT" "I0" ON "A0"."PURL" = "I0"."PURL"
+                LEFT JOIN "INTEGRITY_ANALYSIS" "IA" ON "A0"."ID" = "IA"."COMPONENT_ID"
+                LEFT OUTER JOIN "LICENSE" "D0" ON "A0"."LICENSE_ID" = "D0"."ID"
+                WHERE "A0"."PROJECT_ID" = ?
+                """;
+
         if (onlyOutdated) {
             // Components are considered outdated when metadata does exists, but the version is different than latestVersion
             // Different should always mean version < latestVersion
             // Hack JDO using % instead of .* to get the SQL LIKE clause working:
-            querySring +=
-                    " && !(" +
-                            " SELECT FROM org.dependencytrack.model.RepositoryMetaComponent m " +
-                            " WHERE m.name == this.name " +
-                            " && m.namespace == this.group " +
-                            " && m.latestVersion != this.version " +
-                            " && this.purl.matches('pkg:' + m.repositoryType.toString().toLowerCase() + '/%') " +
-                            " ).isEmpty()";
+            queryString +=
+                    """
+                        AND NOT (NOT EXISTS (
+                        SELECT "M"."ID"
+                        FROM "REPOSITORY_META_COMPONENT" "M" WHERE "M"."NAME" = "A0"."NAME"
+                        AND "M"."NAMESPACE" = "A0"."GROUP"
+                        AND "M"."LATEST_VERSION" <> "A0"."VERSION"
+                        AND "A0"."PURL" LIKE (('pkg:' || LOWER("M"."REPOSITORY_TYPE")) || '/%') ESCAPE E'\\\\'))
+                    """;
         }
         if (onlyDirect) {
-            querySring +=
-                    " && this.project.directDependencies.matches('%\"uuid\":\"'+this.uuid+'\"%') "; // only direct dependencies
+            queryString +=
+                    """
+                       AND "B0"."DIRECT_DEPENDENCIES" LIKE (('%' || "A0"."UUID") || '%') ESCAPE E'\\\\'
+                    """;
         }
-        final Query<Component> query = pm.newQuery(querySring);
-        query.getFetchPlan().setMaxFetchDepth(2);
         if (orderBy == null) {
-            query.setOrdering("name asc, version desc, id asc");
-        }
-        if (filter != null) {
-            final String filterString = ".*" + filter.toLowerCase() + ".*";
-            result = execute(query, project, filterString);
+            queryString +=
+                    """
+                        ORDER BY "name",
+                        "version" DESC
+                    """;
         } else {
-            result = execute(query, project);
+            if (orderBy.equalsIgnoreCase("componentMetaInformation.publishedDate")) {
```

Comment:
would it be more readable to use switch expression on the orderBy value and then have the ORDER BY part appended to the query with the final result?